File: dotnet/Erp.ErrorManagement.Core/Fingerprint.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.RegularExpressions;

namespace Erp.ErrorManagement
{
    /// <summary>
    /// The .NET half of the deduplication algorithm.
    ///
    /// This MUST produce the same hash as fingerprint.ts for the same logical
    /// error, otherwise an Angular-reported and an API-reported instance of one
    /// fault open two fingerprints, two tickets and two counts.  The rules are
    /// documented once in docs/ARCHITECTURE.md §6 and implemented twice, here
    /// and there; tools/verify-fingerprint.mjs and FingerprintParityTests assert
    /// the two agree on a shared corpus of signatures.
    /// </summary>
    public static class Fingerprint
    {
        private const RegexOptions Opts = RegexOptions.Compiled | RegexOptions.CultureInvariant;

        private static readonly Regex RxGuid = new Regex(
            @"\b[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}\b", Opts);
        private static readonly Regex RxIsoDate = new Regex(
            @"\b\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2}(:\d{2}(\.\d+)?)?(Z|[+-]\d{2}:?\d{2})?)?\b", Opts);
        private static readonly Regex RxQuotedSingle = new Regex(@"'[^']*'", Opts);
        private static readonly Regex RxQuotedDouble = new Regex(@"""[^""]*""", Opts);
        private static readonly Regex RxHex = new Regex(@"\b0x[0-9a-fA-F]+\b", Opts);
        private static readonly Regex RxLongHex = new Regex(@"\b[0-9a-fA-F]{16,}\b", Opts);
        private static readonly Regex RxNumber = new Regex(@"\b\d+(\.\d+)?\b", Opts);
        private static readonly Regex RxUrl = new Regex(@"\bhttps?://[^\s)'""]+", Opts);
        private static readonly Regex RxWindowsPath = new Regex(@"\b[a-zA-Z]:\\[^\s:)'""]+", Opts);
        private static readonly Regex RxEmail = new Regex(@"\b[^\s@]+@[^\s@]+\.[^\s@]+\b", Opts);
        private static readonly Regex RxWs = new Regex(@"\s+", Opts);

        /// <summary>Reduce a message to its shape.  Order matters - see the TS twin.</summary>
        public static string NormalizeMessage(string message)
        {
            if (string.IsNullOrEmpty(message)) return string.Empty;

            var s = message;
            s = RxUrl.Replace(s, "{url}");
            s = RxWindowsPath.Replace(s, "{path}");
            s = RxEmail.Replace(s, "{email}");
            s = RxGuid.Replace(s, "{guid}");
            s = RxIsoDate.Replace(s, "{date}");
            s = RxQuotedSingle.Replace(s, "'{str}'");
            s = RxQuotedDouble.Replace(s, "\"{str}\"");
            s = RxHex.Replace(s, "{hex}");
            s = RxLongHex.Replace(s, "{hex}");
            s = RxNumber.Replace(s, "{n}");
            s = RxWs.Replace(s, " ").Trim();

            return s.Length > 500 ? s.Substring(0, 500) : s;
        }

        /// <summary>
        /// Keep the call path, drop the file, line and column.
        ///
        /// A .NET stack frame reads "at Erp.Sales.OrderService.Post(Int32 id) in
        /// C:\build\src\OrderService.cs:line 88".  Everything from " in " is
        /// build-machine detail that changes with every release; the method
        /// signature is the part that identifies the fault.
        /// </summary>
        public static List<string> NormalizeStackFrames(string stackTrace, int depth = 5)
        {
            var frames = new List<string>();
            if (string.IsNullOrEmpty(stackTrace)) return frames;

            foreach (var rawLine in stackTrace.Split('\n'))
            {
                var line = rawLine.Trim();
                if (line.Length == 0) continue;

                if (line.StartsWith("at ", StringComparison.Ordinal)) line = line.Substring(3);
                else if (line.StartsWith("--- End of", StringComparison.Ordinal)) continue;
                else if (!line.Contains("(")) continue;

                var inIndex = line.IndexOf(" in ", StringComparison.Ordinal);
                if (inIndex > 0) line = line.Substring(0, inIndex);

                // Drop the parameter list: an overload resolved differently
                // between builds is still the same fault.
                var paren = line.IndexOf('(');
                if (paren > 0) line = line.Substring(0, paren);

                line = line.Trim();
                if (line.Length == 0) continue;

                // Framework plumbing is identical for every fault and dilutes
                // the signature - the same reason zone.js frames are dropped
                // on the Angular side.
                if (line.StartsWith("System.Runtime.CompilerServices", StringComparison.Ordinal) ||
                    line.StartsWith("System.Threading.Tasks", StringComparison.Ordinal) ||
                    line.StartsWith("System.Runtime.ExceptionServices", StringComparison.Ordinal))
                    continue;

                frames.Add(line);
                if (frames.Count >= depth) break;
            }

            return frames;
        }

        /// <summary>'/api/invoices/4821/lines?page=2' -> '/api/invoices/{id}/lines'</summary>
        public static string NormalizeEndpoint(string endpoint)
        {
            if (string.IsNullOrEmpty(endpoint)) return string.Empty;

            var s = endpoint.Split('?')[0];
            s = RxGuid.Replace(s, "{id}");
            s = Regex.Replace(s, @"/\d+(?=/|$)", "/{id}");
            s = s.TrimEnd('/');
            return s.ToLowerInvariant();
        }

        public class Input
        {
            public string Layer { get; set; }
            public string Category { get; set; }
            public string ExceptionType { get; set; }
            public string Message { get; set; }
            public string StackTrace { get; set; }
            public string Component { get; set; }
            public string Screen { get; set; }
            public string ErpModule { get; set; }
            public string ApiController { get; set; }
            public string ApiAction { get; set; }
            public string ApiEndpoint { get; set; }
            public int? HttpStatusCode { get; set; }
            public int? SqlErrorNumber { get; set; }
            public string SqlObjectName { get; set; }
            public string FormName { get; set; }
            public string LovName { get; set; }
            public IEnumerable<string> ValidationKeys { get; set; }
            public int StackFrameDepth { get; set; } = 5;
        }

        public class Result
        {
            public string Hash { get; set; }
            public string Signature { get; set; }
            public string NormalizedMessage { get; set; }
        }

        private static IEnumerable<string> LocationParts(Input i)
        {
            switch (i.Layer)
            {
                case ErrorLayers.Angular:
                    return new[] { i.ErpModule ?? "", i.Component ?? i.Screen ?? "", i.FormName ?? "", i.LovName ?? "" };
                case ErrorLayers.Http:
                    return new[]
                    {
                        NormalizeEndpoint(i.ApiEndpoint),
                        i.HttpStatusCode.HasValue
                            ? i.HttpStatusCode.Value.ToString(CultureInfo.InvariantCulture)
                            : ""
                    };
                case ErrorLayers.WebApi:
                case ErrorLayers.Business:
                case ErrorLayers.Data:
                    return new[] { i.ApiController ?? "", i.ApiAction ?? "" };
                case ErrorLayers.Database:
                    return new[]
                    {
                        i.SqlErrorNumber.HasValue
                            ? i.SqlErrorNumber.Value.ToString(CultureInfo.InvariantCulture)
                            : "",
                        i.SqlObjectName ?? ""
                    };
                default:
                    return new[] { i.ErpModule ?? "", i.Screen ?? "" };
            }
        }

        public static Result Compute(Input input)
        {
            var normalizedMessage = NormalizeMessage(input.Message);
            var frames = NormalizeStackFrames(input.StackTrace, input.StackFrameDepth);

            var parts = new List<string>
            {
                input.Layer ?? "",
                input.Category ?? "",
                (input.ExceptionType ?? "").Trim(),
                normalizedMessage
            };
            parts.AddRange(LocationParts(input));

            if (input.ValidationKeys != null)
            {
                // Ordinal sort, to match JavaScript's default Array.sort() -
                // which compares UTF-16 code units, NOT the current culture.
                // StringComparer.Ordinal is the equivalent; the default
                // Array.Sort would disagree on any non-ASCII control name.
                parts.AddRange(input.ValidationKeys.OrderBy(k => k, StringComparer.Ordinal));
            }

            parts.AddRange(frames);

            var signature = string.Join("|", parts.Select(p => (p ?? "").Replace("|", "/")));

            return new Result
            {
                Hash = Sha256Hex(signature),
                Signature = signature.Length > 1000 ? signature.Substring(0, 1000) : signature,
                NormalizedMessage = normalizedMessage
            };
        }

        public static string Sha256Hex(string input)
        {
            using (var sha = SHA256.Create())
            {
                var bytes = sha.ComputeHash(Encoding.UTF8.GetBytes(input ?? string.Empty));
                var sb = new StringBuilder(64);
                foreach (var b in bytes) sb.Append(b.ToString("x2", CultureInfo.InvariantCulture));
                return sb.ToString();
            }
        }
    }
}

File: angular/erp-error-workspace/projects/erp-error-management/src/lib/core/fingerprint.ts
import { sha256Hex } from './sha256';
import { ErpErrorCategory, ErpErrorLayer } from '../models/error-envelope';

/**
 * Deduplication.
 *
 * The problem the brief describes - "repeated occurrences of the same or
 * substantially similar error should be grouped without creating duplicate
 * tickets" - is entirely a question of what counts as "the same".  Get it too
 * loose and two unrelated faults share one ticket; too tight and one broken LOV
 * opens 4,000 tickets because each carries a different record id in its message.
 *
 * The approach here: build a SIGNATURE from only the stable parts of an error,
 * hash it, and treat the hash as the identity of the problem.
 *
 * Stable:    layer, category, exception type, the SHAPE of the message,
 *            the top few stack frames by function name, and the location
 *            (component / endpoint / SQL object) appropriate to the layer.
 * Volatile:  ids, GUIDs, dates, quoted literals, numbers, file paths, line and
 *            column numbers, minified bundle hashes, user names, timestamps.
 *
 * The signature text is stored next to the hash in ERM.ERM_ErrorFingerprint, so
 * an administrator can always see WHY two errors were grouped - a hash nobody
 * can explain is a hash nobody will trust.
 *
 * The identical algorithm exists in C# (Erp.ErrorManagement.Core/Fingerprint.cs)
 * so an Angular-reported and a .NET-reported instance of the same fault land on
 * the same row.
 */

const RX_GUID = /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/gi;
const RX_ISO_DATE = /\b\d{4}-\d{2}-\d{2}([T ]\d{2}:\d{2}(:\d{2}(\.\d+)?)?(Z|[+-]\d{2}:?\d{2})?)?\b/g;
const RX_QUOTED_SINGLE = /'[^']*'/g;
const RX_QUOTED_DOUBLE = /"[^"]*"/g;
const RX_HEX = /\b0x[0-9a-f]+\b/gi;
const RX_LONG_HEX = /\b[0-9a-f]{16,}\b/gi;
const RX_NUMBER = /\b\d+(\.\d+)?\b/g;
const RX_URL = /\bhttps?:\/\/[^\s)'"]+/gi;
const RX_WINDOWS_PATH = /\b[a-z]:\\[^\s:)'"]+/gi;
const RX_EMAIL = /\b[^\s@]+@[^\s@]+\.[^\s@]+\b/g;
const RX_WS = /\s+/g;

/**
 * Reduce a message to its shape.  Order matters: URLs and paths are replaced
 * before numbers, otherwise the port and the path segments become {n}{n}{n} and
 * two different endpoints collapse into one.
 */
export function normalizeMessage(message: string | null | undefined): string {
  if (!message) return '';
  return message
    .replace(RX_URL, '{url}')
    .replace(RX_WINDOWS_PATH, '{path}')
    .replace(RX_EMAIL, '{email}')
    .replace(RX_GUID, '{guid}')
    .replace(RX_ISO_DATE, '{date}')
    .replace(RX_QUOTED_SINGLE, "'{str}'")
    .replace(RX_QUOTED_DOUBLE, '"{str}"')
    .replace(RX_HEX, '{hex}')
    .replace(RX_LONG_HEX, '{hex}')
    .replace(RX_NUMBER, '{n}')
    .replace(RX_WS, ' ')
    .trim()
    .slice(0, 500);
}

/**
 * Keep the call path, drop everything about WHERE the file lived.
 *
 * A production Angular bundle changes its chunk hash on every deploy, and the
 * line numbers move with any edit above the fault.  Including either of those
 * means the same bug fingerprints differently after every release, the
 * occurrence count resets to 1, and the recurring-problem report - the thing
 * that makes a permanent fix worth doing - never fires.
 */
export function normalizeStackFrames(stack: string | null | undefined, depth = 5): string[] {
  if (!stack) return [];
  const frames: string[] = [];

  for (const rawLine of stack.split('\n')) {
    const line = rawLine.trim();
    if (!line || line === 'Error' || /^[A-Za-z.]*(Error|Exception):/.test(line)) continue;

    let frame = line
      // Chrome/Edge: "at ClassName.method (http://host/main-ABC123.js:1:2345)"
      .replace(/^at\s+/i, '')
      // Firefox/Safari: "method@http://host/main.js:1:2345"
      .replace(/@.*$/, '')
      // Strip the parenthesised location entirely.
      .replace(/\s*\(.*\)\s*$/, '')
      // Any surviving bare location.
      .replace(/https?:\/\/\S+/g, '')
      .replace(/:\d+:\d+$/, '')
      .trim();

    // Webpack/Vite name-mangled anonymous frames carry no information.
    if (!frame || frame === '<anonymous>' || frame === 'Object.<anonymous>') continue;
    // Framework frames are the same for every fault; they dilute the signature.
    if (/^(zone|Zone|ZoneDelegate|ZoneTask|invokeTask|runTask|drainMicroTaskQueue)\b/.test(frame)) continue;

    frames.push(frame);
    if (frames.length >= depth) break;
  }

  return frames;
}

export interface FingerprintInput {
  layer: ErpErrorLayer;
  category: ErpErrorCategory;
  exceptionType?: string | null;
  message?: string | null;
  stackTrace?: string | null;
  /** Angular layer. */
  component?: string | null;
  screen?: string | null;
  erpModule?: string | null;
  /** HTTP / API layers. */
  apiController?: string | null;
  apiAction?: string | null;
  apiEndpoint?: string | null;
  httpStatusCode?: number | null;
  /** Database layer. */
  sqlErrorNumber?: number | null;
  sqlObjectName?: string | null;
  /** Form / LOV layers. */
  formName?: string | null;
  lovName?: string | null;
  /** Validation errors contribute their control+rule pairs, never their values. */
  validationKeys?: string[] | null;
  stackFrameDepth?: number;
}

export interface FingerprintResult {
  hash: string;
  signature: string;
  normalizedMessage: string;
}

/**
 * Which location fields belong in the signature depends on the layer.  A SQL
 * deadlock in usp_PostJournal is the same problem whichever screen triggered
 * it; an Angular template error in InvoiceLineComponent is not the same problem
 * as one in CustomerSearchComponent even if the message is identical.
 */
function locationParts(i: FingerprintInput): string[] {
  switch (i.layer) {
    case 'angular':
      return [i.erpModule ?? '', i.component ?? i.screen ?? '', i.formName ?? '', i.lovName ?? ''];
    case 'http':
      // Endpoint without its query string or embedded ids - those are volatile.
      return [normalizeEndpoint(i.apiEndpoint), String(i.httpStatusCode ?? '')];
    case 'webapi':
    case 'business':
    case 'data':
      return [i.apiController ?? '', i.apiAction ?? ''];
    case 'database':
      return [String(i.sqlErrorNumber ?? ''), i.sqlObjectName ?? ''];
    default:
      return [i.erpModule ?? '', i.screen ?? ''];
  }
}

/** '/api/invoices/4821/lines?page=2' -> '/api/invoices/{id}/lines' */
export function normalizeEndpoint(endpoint: string | null | undefined): string {
  if (!endpoint) return '';
  return endpoint
    .split('?')[0]
    .replace(RX_GUID, '{id}')
    .replace(/\/\d+(?=\/|$)/g, '/{id}')
    .replace(/\/+$/, '')
    .toLowerCase();
}

export function computeFingerprint(input: FingerprintInput): FingerprintResult {
  const normalizedMessage = normalizeMessage(input.message);
  const frames = normalizeStackFrames(input.stackTrace, input.stackFrameDepth ?? 5);

  const parts = [
    input.layer,
    input.category,
    (input.exceptionType ?? '').trim(),
    normalizedMessage,
    ...locationParts(input),
    ...(input.validationKeys ?? []).slice().sort(),
    ...frames,
  ];

  // '|' is the separator; strip it from the parts so a value containing a pipe
  // cannot be crafted to collide with a different part layout.
  const signature = parts.map((p) => String(p ?? '').replace(/\|/g, '/')).join('|');

  return { hash: sha256Hex(signature), signature: signature.slice(0, 1000), normalizedMessage };
}
